lib/rubocop/cop/packaging/relative_require_to_lib.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module RuboCop # :nodoc:
+  module Cop # :nodoc:
+    module Packaging # :nodoc:
+      # TODO
+      # Write about this cop.
+      # Write bad and good methods.
+      # Help from cop/lint/redundant_require_statement.rb.
+      #
+      class RelativeRequireToLib < Cop
+        # This is the message that will be displayed when RuboCop finds an
+        # offense of using `require` with relative path to lib.
+        MSG = 'Avoid using `require` with relative path to lib.'
+
+        def_node_matcher :require?, <<~PATTERN
+          {(send nil? :require (str #starts_with_relative_path?))
+           (send nil? :require (send (const nil? :File) :expand_path (str #starts_with_relative_path?) {(str _) (send nil? _)}))
+           (send nil? :require (send (send (const nil? :File) :dirname {(str _) (send nil? _)}) :+ (str #starts_with_relative_path?)))}
+        PATTERN
+
+        # Extended method.
+        # TODO: document it.
+        def on_send(node)
+          return unless require?(node)
+
+          add_offense(node)
+        end
+
+        # This method is called from inside `#def_node_search`.
+        # It is used to find the strings that uses relative path.
+        def starts_with_relative_path?(str)
+          str.match?(%r{.*\./lib/})
+        end
```

Comment:
Hi @terceiro, @deivid-rodriguez,
This cop works alright now, with this regex :P
But since we discussed to not use regex and use paths to determine if the `require` call is to the `lib` directory, I could possibly come up with this diff:

```
--- a/lib/rubocop/cop/packaging/relative_require_to_lib.rb
+++ b/lib/rubocop/cop/packaging/relative_require_to_lib.rb
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'pathname'
+
 module RuboCop # :nodoc:
   module Cop # :nodoc:
     module Packaging # :nodoc:
@@ -30,7 +32,12 @@ module RuboCop # :nodoc:
         # This method is called from inside `#def_node_search`.
         # It is used to find the strings that uses relative path.
         def starts_with_relative_path?(str)
-          str.match?(%r{.*\./lib/})
+          current_dir = Dir.pwd
+          relative_dir = File.expand_path(str)
+          relative_dir.delete_prefix!(current_dir + "/")
+          relative_dir.start_with?('lib')
         end
       end
     end
```

Does this look okay? I am not sure how "professional" this is, I just tried to make it work in the best possible way first :sweat_smile: 
Please let me know what you think!?

Reply:
`current_dir` needs to be the root directory of the project being analyzed, not `Dir.pwd`. It happens to be the same as `Dir.pwd` if you are running rubocop from the root, but that's not always the case.

Reply:
Ah, right. My bad :man_facepalming: 
So my idea then is to have some sort of `config.rb` or `environment.rb` where I define the `ROOT` constant as `__dir__` and then use it here?

Reply:
Hm, I guess something like this should do the trick?
```
--- a/lib/rubocop/cop/packaging/relative_require_to_lib.rb
+++ b/lib/rubocop/cop/packaging/relative_require_to_lib.rb
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require_relative '../../packaging/version'
+
 module RuboCop # :nodoc:
   module Cop # :nodoc:
     module Packaging # :nodoc:
@@ -30,7 +32,10 @@ module RuboCop # :nodoc:
         # This method is called from inside `#def_node_search`.
         # It is used to find the strings that uses relative path.
         def starts_with_relative_path?(str)
-          str.match?(%r{.*\./lib/})
+          root_dir = RuboCop::Packaging::ROOT
+          relative_dir = File.expand_path(str)
+          relative_dir.delete_prefix!(root_dir + "/")
+          relative_dir.start_with?('lib')
         end
       end
     end
diff --git a/lib/rubocop/packaging/version.rb b/lib/rubocop/packaging/version.rb
index 96af894..eba05a6 100644
--- a/lib/rubocop/packaging/version.rb
+++ b/lib/rubocop/packaging/version.rb
@@ -3,5 +3,6 @@
 module RuboCop
   module Packaging
     VERSION = '0.1.0'
+    ROOT = File.expand_path('../../..', __dir__)
   end
 end
```

P.S. I am just experimenting to see if this fits the requirement we have. Don't mind that I added the `ROOT` constant in the `version.rb` file for now :see_no_evil:

Reply:
Oh crap. I just realized how stupid this is! :man_facepalming::man_facepalming::man_facepalming::man_facepalming::man_facepalming: 
(or I think I am horribly confused!)